### PR TITLE
Add "Why mocking fails?" post

### PR DESCRIPTION
**Pull request by nelsonsar**
A post about the purpose of mocks and the guideline that can make you avoid mock hell.

### CODE REVIEW

**Comment by nelsonsar** [comment]
It needs more references:

1) GoF: programming to an interface
2) Invariants

File: 2016-10-16-why-mocking-fails [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Why mocking fails?
+
+[Yegor argues that fake objects are better than mocks](http://www.yegor256.com/2014/09/23/built-in-fake-objects.html) shows that mocks can be really hard to build when objects create another objects and these another objects create more objects and so on (maybe forever?) and I agree with it (this is the beginning of a **mock hell**).
```
Comment (augustohp) [inline 2016-10-16-why-mocking-fails]: "Yegor argument"

**Review by augustohp — APPROVED** [review]
😍

File: 2016-10-16-why-mocking-fails.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Why mocking fails?
+
+[Yegor arguments that fake objects are better than mocks](http://www.yegor256.com/2014/09/23/built-in-fake-objects.html) shows that mocks can be really hard to build when objects create another objects and these another objects create more objects and so on (maybe forever?) and I agree with it (this is the beginning of a **mock hell**).
```
Comment (marcvitbr) [inline 2016-10-16-why-mocking-fails.md:3]: The plural makes it tricky, but I think the correct form would be:

"_**Yegor's** arguments..._"

File: 2016-10-16-why-mocking-fails.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+But, the whole point is: should we create all these mocks returning mocks? Or should we rely on the fake objects the post suggests?
+
+## The great virtue of OOP
+
+[John Cinnamond](https://twitter.com/jcinnamond) in a recent [talk](http://brightonruby.com/2016/the-point-of-objects-john-cinnamond/) makes a brilliant conclusion about OOP: it's a tool used to desconstruct a complex problem and defer decisions (implementations). By using it you (your team, your company, whatever) can focus in one problem at a time.
```
Comment (marcvitbr) [inline 2016-10-16-why-mocking-fails.md:9]: **deconstruct** instead of de**s**construct.

File: 2016-10-16-why-mocking-fails.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+In the example above you don't have any idea of what `region` actually is or how it gets the salary information, I just describe how the interaction occurs.
+
+For me the reason to use mocks is: represent the interaction between objects while [programming for an interface not for an implementation](http://www.artima.com/lejava/articles/designprinciples.html).
+
+Programming to an interface enables the replacement of the `region` dependency with anything that responds to the `salary_by` message (I can even create a fake class, to hide `region` implementation, using a `Hash` for example) and defer the decision of which ORM, database or library I'd like to use to another team or just do it later.
+
+## You can only defer what you can control
+
+But, in order to enable mocks hidden powers (or make them shine) you need to control them and that's where mocks always goes wrong...
```
Comment (marcvitbr) [inline 2016-10-16-why-mocking-fails.md:42]: Another tricky one involving plurals :)
My suggestions are:

_"mocks always **go** wrong"_ or _"**mocking** always goes wrong"_

File: 2016-10-16-why-mocking-fails.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+The [Mock Roles, not Objects paper](http://www.jmock.org/oopsla2004.pdf) stabilish a set of simple rules (ok, not rules... guidelines) for mocks and the most important one (and ignored most of the time too), IMO, is: "Only Mock Types You Own". Following this simple guideline you'll **always** have to mock an interface created by you, which is most of the time simpler than the one exposed by a library (because of the level of abstraction it needs).
+
+**Simpler interfaces == easier to test and mantain** (less public methods, less coupling and higher cohesion).
+
+So, if you find yourself creating a mock that returns other mock or mocking a third party library think twice, be sure that the operation shouldn't be better encapsulated (maybe is a missing domain aspect in your app, for example) and don't be afraid of changing it if you think a mock hell is knocking at your door (or can emerge soon).
```
Comment (marcvitbr) [inline 2016-10-16-why-mocking-fails.md:48]: _"third**-**party"_ instead of _"third party"_

**Review by marcvitbr — COMMENTED** [review]
Overall, I think the post is a very good reminder of the best practices for mocks, with a ton of awesome references.

I specially like how you bring up the decision-referring aspect of mocking, which highlights the design.

Anyways, I've just made a couple of (totally dispensable) suggestions regarding the grammar. :)

**Pull request merged**